### PR TITLE
Enable Aggregate Operation

### PR DESCRIPTION
**Pull request by kidrahahjo**
## Description
Enable the use of aggregate operation in `flow`. 

#### How will jobs get aggregated?
- For operations, jobs will be aggregated based on aggregators associated with operations.
- For group, jobs will be aggregated based on aggregators associated with groups. 

#### Aggregate IDs
Similar to the concept of job ids, an aggregate id is a unique id assigned to every aggregate which is created in a project. An aggregate id will always have a prefix `agg-` so that users can distinguish between aggregate id and a job id.
Users can calculate the aggregate id of a job using flow's Python API.
``` python
from flow import get_aggregate_id
project = # Code to get a signac project
list_of_jobs = list(project)
id = get_aggregate_id(project)
print(id)  # Prints aggregate id for an aggregate of all jobs in a project.
```
#### Ways of Aggregation
- Aggregation of all jobs in a project.
- Aggregation of jobs grouped by keys, where keys can be state points, callables, or a sequence of state points. 
- Aggregation of jobs in "groups of" fashion.
- Sorting, or selecting jobs in an aggregate based on state points or a condition.

#### Code Example
``` python
# project.py

from flow import FlowProject, aggregator

group = FlowProject.make_group('agg_group', aggregator_obj=aggregator())

@FlowProject.operation
@aggregator()
@group
def op_1(*jobs):
    pass

@FlowProject.operation
@group
def op_2(*jobs): 
    pass
```

``` bash
# Run operations op_1 and op_2 individually. 
# op_1 acts as an aggregate operation, op_2 acts as a non-aggregate operation
$ python project.py run -o op_1 op_2

# Run operations op_1 and op_2 as a group.
# Both op_1 and op_2 acts as aggregate operations
$ python project.py run -o agg_group

# Run operation op_1 for an aggregate whose aggregate id is `agg-asdg3425jgnoet`
# op_1 will only get executed if the aggregator associated will op_1 generated an aggregate whose aggregate id is `agg-asdg3425jgnoet`
$ python project.py run -o agg_group -j agg-asdg3425jgnoet
```

## Motivation and Context

Explained in detail above!

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [ ] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

**Comment by vyasr** [comment]
@kidrahahjo I recommend that you take the opportunity now to open a new PR in the [signac-docs repo](https://github.com/glotzerlab/signac-docs/) that documents the aggregation feature. This is probably the best time for you to do it since all the logic will be fresh and you'll already be writing relevant explanations as part of this thread.

Is this PR waiting on anything else at this point (other issues being resolved, etc) or do you think that once this PR is approved we'll just be ready?

**Comment by kidrahahjo** [comment]
@vyasr That's a good suggestion, I will start working on adding documentation for aggregation. For the second point of yours, I did not write regression tests for submission in order to restrict the length of this PR. Once this gets ready to merge, I will write submission tests and we can merge that pull request into this one. What do you think?

As far as documentation is concerned, @vyasr @bdice do you want me to write the documentation in a glossary like fashion? For more context, in yesterday's office hours, we came to a point that our concept page should be glossary like (inspired by what Airflow has on their [concepts page](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html)). For topic documentation (like aggregation, groups, etc), we can choose to document things in a glossary like fashion as well.

File: flow/project.py
Comment (kidrahahjo) [inline flow/project.py:3566]:
```suggestion
        jobs : iterable of :class:`~signac.contrib.job.Job` or iterable of aggregate of :class:`~signac.contrib.job.Job`
```

File: flow/project.py
Comment (kidrahahjo) [inline flow/project.py:3568]:
```suggestion
            or all the associated aggregates/jobs for an operation if the argument is None. (Default value = None)
```

This document really raises the question that do we want our users to know that internally, a normal operation always acts on aggregate of single job?

File: flow/project.py
Comment (kidrahahjo) [inline flow/project.py:4214]:
```suggestion
            An instance of :class:`~flow.aggregator` to associate with the:class:`FlowGroup`. If None, the
```

aggregator object seems like a vague term, this feels more clear.

File: flow/project.py
Comment (bdice) [reply flow/project.py:3568]: I think the suggestion is okay but not necessary. Either way is fine with me. Maybe get another opinion on this.

File: flow/project.py
Comment (bdice) [reply flow/project.py:4214]: Seems fine to me.

File: flow/project.py
Comment (bdice) [inline flow/project.py:4254]: I'm confused, is this comment related to the comment above? (quoted here)

```
        # We register aggregators associated with operation functions
        # in `_register_groups` and hence the aggregator parameter is passed as None.
        # This is because we do not restrict the decorator placement in terms of
        # `FlowGroupEntry`, `aggregator`, or `operation`.
```

File: flow/project.py
Comment (bdice) [inline flow/project.py:4291]: Is it possible to combine this logic with the original registration, so that there is only one implementation used by "registration" and "re-registration"?

File: flow/project.py
Comment (bdice) [inline flow/project.py:4546]:
```suggestion
                if id.startswith("agg-"):
```

File: flow/project.py
Comment (bdice) [inline flow/project.py:4545]: Change this variable name (perhaps `job_id`) so that it doesn't shadow the built-in function `id`.

File: flow/project.py
Comment (bdice) [inline flow/project.py:4780]:
```suggestion
            help="The job ids or aggregate ids in the FlowProject. "
```

File: flow/templates/base_status.jinja
Comment (bdice) [reply flow/templates/base_status.jinja:31]: This is related to previous discussions where we've considered names like `ExecutionUnit` or `SubmissionTask`, we need to create a separate issue for this because we do need a name for that unit of work.

**Review by bdice — CHANGES_REQUESTED** [review]
@kidrahahjo Thanks for your work on this! I have some comments, but I think I will need to come back to it with fresh eyes for a better review.

File: flow/templates/base_status.jinja
Comment (kidrahahjo) [reply flow/templates/base_status.jinja:31]: Then I believe that we should not make this change in this PR. I'll revert the related changes.

File: flow/project.py
Comment (kidrahahjo) [reply flow/project.py:178]: I agree, let me add this to #362 as a task for aggregation.

File: flow/project.py
Comment (kidrahahjo) [reply flow/project.py:4254]: Yes, this is related to the quoted comment. I think of it as a hack to disable to placement requirements of the `aggregator` decorator.

File: flow/project.py
Comment (kidrahahjo) [reply flow/project.py:702]: This is not quite true, `None` doesn't necessarily mean that the default value is used. When used for a `FlowGroup` (created using `FlowProject.make_group`), only then `None` would mean the default aggregation is used. See this comment for reference

```
# We register aggregators associated with operation functions
# in `_register_groups` and hence the aggregator parameter is passed as None.
# This is because we do not restrict the decorator placement in terms of
# `FlowGroupEntry`, `aggregator`, or `operation`.
```

File: tests/test_project.py
Comment (bdice) [inline tests/test_project.py:1898]: I think you can do this instead of splitting/joining. I recently made this change in some other tests. Match this change (remove `splittlines` and `join`) in the other tests below, please.
```suggestion
        run_output = self.call_subcmd("run -o agg_op4").decode("utf-8")

        assert "1 and 2" in run_output
```

**Review by bdice — CHANGES_REQUESTED** [review]
Some additional comments.

File: flow/project.py
Comment (vyasr) [reply flow/project.py:4256]: Can you remind me why exactly this function is necessary to make part of the API again? Is it because the user needs to tell flow when jobs have been added to the data space? The implementation is different enough from the way that aggregators are determined inside `_register_groups` to give me a little pause.

File: tests/define_test_aggregate_project.py
Comment (vyasr) [inline tests/define_test_aggregate_project.py:8]: This name is too generic and will make it harder for readers of the tests to know what the custom aggregation is based on. Rename it to something like `sp_i_parity_aggregator`.

File: tests/define_test_aggregate_project.py
Comment (vyasr) [inline tests/define_test_aggregate_project.py:4]: There's no need to make this internal. This is part of our testing suite, not the package.
```suggestion
class TestAggregateProject(FlowProject):
```

File: tests/define_test_aggregate_project.py
Comment (vyasr) [reply tests/define_test_aggregate_project.py:8]: On second thought, I guess `parity` may not be a word everyone's familiar with, I'm not sure. If the next reviewer who sees this thinks that might be the case, then maybe we use an even simpler name like `sp_i_even_odd_aggregator`. Either name is fine with me, although my first suggestion is a bit less verbose :)

File: tests/define_test_aggregate_project.py
Comment (vyasr) [inline tests/define_test_aggregate_project.py:27]: All your tests are doing this. I'd define a small function
```
def set_all_job_docs(jobs, key, val):
    for job in jobs:
        job[key] = val
```
and use that. Then half of these operations become trivial (setting `True`) and the others do one calculation and call this.

File: tests/define_test_aggregate_project.py
Comment (vyasr) [inline tests/define_test_aggregate_project.py:69]:
```suggestion
# This operation should raise an exception and only exists to test that.
@_TestAggregateProject.operation
```

File: tests/test_project.py
Comment (vyasr) [inline tests/test_project.py:1830]: I don't think I understand what you're trying to say here. The "default aggregate" is just an individual job, right?

**Review by vyasr — CHANGES_REQUESTED** [review]

**Comment by bdice** [comment]
I'm going to spend a little time fixing up open review comments.

File: tests/define_test_aggregate_project.py
Comment (bdice) [reply tests/define_test_aggregate_project.py:4]: This should not be changed, it matched the style of the other projects in our testing suite. However, it should be named like:
```suggestion
class _AggregateTestProject(FlowProject):
```

Comment (bdice) [reply tests/define_test_aggregate_project.py]: Resolved in 7a8a806d52a8c6a01a8f32db90d087aa2e32b1fb.

File: tests/define_test_aggregate_project.py
Comment (bdice) [reply tests/define_test_aggregate_project.py]: Resolved in 1f3efb6b4ff57c8e9094347e3bc9a5a71ca22251.

File: flow/project.py
Comment (bdice) [reply flow/project.py]: Renamed to `group_aggregator` in 76c06048da93724d819de21e4c406241a668b38f, changed elsewhere for consistency and to avoid shadowing the imported name `aggregator`.

File: flow/project.py
Comment (bdice) [reply flow/project.py:4222]: I don't know what to do about this immediately, but I think that it might be best to merge this as-is and simplify this logic later once we remove the legacy `add_operation` feature and consolidate group registration. Related discussion on PRs: #382, #487.

File: flow/render_status.py
Comment (bdice) [reply flow/render_status.py:172]: Bulleted lists require a blank line before and after. Also if the string must be one of two options, you can specify it like the argument "engine" to `pandas.read_csv`:
https://github.com/pandas-dev/pandas/blob/f2c8480af2f25efdbd803218b9d87980f416563e/pandas/io/parsers.py#L176
https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

I'll change that above, edit the docstring, and resolve this comment.

File: flow/render_status.py
Comment (bdice) [inline flow/render_status.py:169]:
```suggestion
        kind : {{"jobs", "aggregates"}}
            Whether to interpret the ``jobs`` argument as individual jobs or aggregates of jobs.
```

File: flow/aggregates.py
Comment (bdice) [reply flow/aggregates.py:445]: I removed this variable initialization from the constructor in 1d40e8309c7152f625d24105468a65bd673468c1.

File: tests/test_project.py
Comment (bdice) [reply tests/test_project.py]: What this means is that all of the operations define a custom aggregator, and thus the aggregates cursor will have some aggregates with the same id. For example, `@aggregator()` and `@aggregator.groupsof(30)` will return the same aggregates (a single aggregate containing all 30 jobs in the project) but the `_AggregatesCursor` is not capable of distinguishing which aggregators are equivalent. Thus, there are 41 aggregates in the cursor, even though there are only 34 unique aggregate ids. I updated the comments and test to clarify this behavior in e8f3701e217df0615c8fc56e9540ca533fc0fbeb.

File: flow/render_status.py
Comment (bdice) [reply flow/render_status.py:161]: The purpose of this template filter is to show an output line in the status overview like:

```
Total # of jobs: 81
Total # of unique aggregates: 27
```

where the project has 81 jobs and 27 "proper" aggregates with more than one job. I am not a huge fan of this additional complexity, but I'm not immediately sure what would be better -- I improved this code in a few ways already and am alright with it for now. A better solution would be to get both values (jobs and aggregates) in a single iteration.

**Comment by bdice** [comment]
I'm finished with the review pass and small updates that I wanted to make for now. There are a handful of significant design questions left to answer in `flow/project.py`, mostly raised by @vyasr. The next step on this PR is to tackle those.

File: flow/project.py
Comment (kidrahahjo) [reply flow/project.py:286]: 1. Yes, any filter provided will prevent finding aggregates.
2. As far as I remember, I think we decided that these filters should not be limited to jobs and hence, we should also be able to apply aggregate filters in future.

File: tests/test_project.py
Comment (b-butler) [inline tests/test_project.py:1888]: Can we not detect this earlier than trying to run the operation?

**Review by b-butler — COMMENTED** [review]
I looked over the code before our meeting today, and left just a few comments along the way.

File: flow/project.py
Comment (bdice) [reply flow/project.py:3008]: Consensus of @bdice @atravitz @b-butler @vyasr : This should be left as-is:
```
iterable of :class:`~signac.contrib.job.Job` or iterable of aggregate of :class:`~signac.contrib.job.Job`
```

File: flow/project.py
Comment (bdice) [reply flow/project.py:3548]: Resolved by consensus on a new wording.

File: flow/project.py
Comment (bdice) [reply flow/project.py:303]: @vyasr has an idea for improvement.

File: flow/project.py
Comment (bdice) [reply flow/project.py]: @b-butler Can you take a look at this? `FlowGroupEntry` allows `group_aggregator=None` but `make_group` replaces `group_aggregator=None` with `group_aggregator=aggregator.groupsof(1)`.

File: flow/project.py
Comment (bdice) [reply flow/project.py:4222]: @b-butler Also this comment is related to the above discussion of group/operation initialization.

File: flow/render_status.py
Comment (bdice) [reply flow/render_status.py]: I thought some more about this and I'm pretty sure that we should pre-compute this information in `print_status` if it is needed. Also there were merge conflicts with PR #501 so the existing design of this PR had to be changed. To punt on further debate about the purpose/scope of this function, I removed it in commit 4cc3cba902986f67008f5ec80b8908f21d54ca6a and we can re-visit this at a later time if we wish to change the status output format to distinguish jobs and aggregates. I'm going to resolve this comment thread for now since it doesn't seem like any of us have strong opinions (except that we don't want this to be a blocker for all of aggregation).

Current output: `Overview: 11 jobs/aggregates, 11 jobs/aggregates with eligible operations.`

File: tests/test_project.py
Comment (bdice) [reply tests/test_project.py:1863]: Status output tests will be handled in another PR that @Charlottez112 has started. Tests for status output are currently outside the scope of this PR so I'll resolve this.

File: flow/project.py
Comment (bdice) [reply flow/project.py:293]: Done: fdf7343d25a1a9cd606721a98fbd35e1dce3f307

File: tests/define_test_aggregate_project.py
Comment (bdice) [reply tests/define_test_aggregate_project.py]: Done in 6266b3728188381fc4af60dadd9fc9f074f9dab1.

File: flow/project.py
Comment (vyasr) [reply flow/project.py:307]: Here's my suggestion. Currently most methods of `AggregatesCursor` behave differently based on the exact same criterion, namely whether or not there are filters, but semantically the presence of a filter indicates something more because if we have a filter we only iterate over a `JobsCursor`, except that we wrap the returned jobs in a tuple. That to me indicates that we should have (very small) subclasses to distinguish the two cases, with one case being a true `AggregatesCursor`, whereas the other is very thin wrapper around a `JobsCursor` that returns tuples instead.

File: tests/test_project.py
Comment (bdice) [reply tests/test_project.py:1937]: Fixed in 865277b2a7b65c31bf2fe9b8ac0a183b073cd5b1.

File: flow/operations.py
Comment (bdice) [reply flow/operations.py:55]: Done.

File: flow/project.py
Comment (bdice) [reply flow/project.py]: Discussed with @kidrahahjo: This has to do with delayed initialization. We cannot know at this point whether there is an aggregator affiliated with the operation. We added this comment at another point in the code, and I copied it here in 1680f46f2f5755e053efe0ab243c1f9b045575a7.
```
        # We register aggregators associated with operation functions in
        # `_register_groups` and we do not set the aggregator explicitly.
        # We delay setting the aggregator because we do not restrict the
        # decorator placement in terms of `@FlowGroupEntry`, `@aggregator`, or
        # `@operation`.
```

File: flow/project.py
Comment (bdice) [reply flow/project.py]: I agree this is worth further investigation. For now, we made the method private in e52aa43. I am going to document this in a separate TODO issue.

**Comment by bdice** [comment]
**Reviewers:** I have worked through this PR with @kidrahahjo and addressed a number of issues. All of the remaining comments pertain to the implementation of `_AggregatesCursor`. I would recommend that we perform a final round of review on this PR, ignoring that class, and make a separate pull request (into this PR or into `master` depending on the timing of merge) that addresses the suggestions from @vyasr and others around `_AggregatesCursor`. I will re-request review from all.

**Review by bdice — APPROVED** [review]
I made a few last rounds of changes to documentation and expanded the tests / code coverage. I think this is ready to merge aside from the outstanding issues with `_AggregatesCursor`. I would vote in favor of merging this PR in its current state and leaving that class refactor as a "TODO" to get done before the next release.

File: tests/test_project.py
Comment (b-butler) [inline tests/test_project.py:1898]:
```suggestion
        assert agg_cursor._project is project
```

This is really an identity check isn't it?

**Review by b-butler — APPROVED** [review]
I have a tiny suggestion that can just be committed or resolved, but I generally agree with @bdice and am approving.

File: tests/test_project.py
Comment (bdice) [reply tests/test_project.py:1898]: I agree. Committing.

**Comment by bdice** [comment]
I think this PR is ready to merge! We're punting on the `_AggregatesCursor` redesign for a follow-up PR. @vyasr You have an outstanding review request but we already know your primary set of change requests on the cursor design. Please feel free to follow up if you have additional comments and we can address those in subsequent PRs. https://github.com/glotzerlab/signac-flow/pull/464

**Pull request merged**
